ls1intum/Artemis
# `Development`: Refactor quiz exercise detail page

### Checklist
#### General


#### Client

### Motivation and Context



The implementation of the Quiz Pool Configuration Page will reuse the functionalities of the Quiz Exercise Detail Page. Therefore, The Quiz Exercise Detail Page is refactored so the functionality can be reused later by the Quiz Pool Configuration page.

### Description


1. Move the UI of the list of Quiz Question Cards to a separate component: QuizQuestionListEditComponent
The list of Quiz Question Cards will also be used on Quiz Pool Configuration Page. Therefore, the UI elements are refactored to a new reusable component.
2. Move the UI for importing existing Quiz Questions to a separate component: QuizQuestionListEditExistingComponent
3. Move the checking of Quiz Questions (if it is valid or not) to the isQuizQuestionValid method.
The checking of Quiz Questions will also be used by the Quiz Pool Configuration Page, so the logic is moved to a separate reusable method.
5. Add ViewEncapsulation.None to MultipleChoiceQuestionEditComponent to make the style consistent with DragAndDropQuestionEditComponent and ShortAnswerQuestionEditComponent.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Course

1. Log in to Artemis
2. Navigate to Course Administration
6. Select the Course
7. Navigate to Exercises
8. Create a new Quiz Exercise
9. Verify that adding new Quiz Questions works
10. Add new Quiz Questions, then Save
11. Edit the newly created Quiz Exercise
12. Verify that the Quiz Questions are displayed correctly
13. Import existing Quiz Questions from either another course, exam, or JSON file.
14. Save the exercise
15. Verify that the imported Quiz Questions are saved

### Review Progress






#### Code Review
#### Manual Tests

### Test Coverage



| Class/File | Branch | Line |
|------------|-------:|-----:|
| multiple-choice-question-edit.component.ts | 87.5% | 95.06% |
| quiz-exercise-detail.component.ts | 76.49% | 89.88% |
| quiz-exercise-validation.directive.ts | 81.96% | 93.22% |
| quiz-question-list-edit-existing.component.ts | 80.55% | 96.7% |
| quiz-question-list-edit.component.ts | 100% | 97.05% |
| quiz-manage-util.service.ts | 96.21% | 99.04% |

### Screenshots



https://user-images.githubusercontent.com/10885503/230735986-53187e78-aac8-4552-a872-a42fa7c6d3cf.mov
